File: medium/0007_reverse_integer/solution.cpp
using namespace std;

#include <climits>
#include <cmath>

class Solution {
public:
  int reverse(int x) {
    if (x == -2147483648)
      return 0;

    const int REVERSE_INT_MAX = INT_MAX / 10;

    const int sign = x < 0 ? -1 : 1;
    x = abs(x);

    int reversedX = 0;

    while (x) {
      if (reversedX > REVERSE_INT_MAX)
        return 0;

      reversedX = reversedX * 10 + x % 10;
      x /= 10;
    }

    return sign * reversedX;
  }
};

File: medium/0007_reverse_integer/solution.py
class Solution:
    def reverse(self, x: int) -> int:
        INT_MAX = 2147483647 / 10

        sign = -1 if x < 0 else 1
        x = abs(x)

        reversed_x = 0

        while x:
            if reversed_x > INT_MAX:
                return 0

            reversed_x = reversed_x * 10 + x % 10
            x //= 10

        return sign * reversed_x
